File: 1813-sentence-similarity-iii/1813-sentence-similarity-iii.cpp
class Solution {
public:
    bool areSentencesSimilar(string sentence1, string sentence2) {
        // since we need to add somewhere in between then it means prefix means string from start should match 
        // strings from end should match
        // so we could take pointers to do this
        // if(sentence1.size() < sentence2.size()) {
        //     swap(sentence1, sentence2);
        // }

        // vector<string> sen1, sen2;

        // stringstream ss1(sentence1);
        // string token;

        // while(ss1 >> token) {
        //     sen1.push_back(token);
        // }

        // stringstream ss2(sentence2);
        // while(ss2 >> token) {
        //     sen2.push_back(token);
        // }

        // int i = 0, j = sen1.size() - 1;
        // int k = 0, l = sen2.size() - 1;

        // while(k < sen2.size() && i < sen1.size() && sen2[k] == sen1[i]) {
        //     k++;
        //     i++;
        // }

        // while(l >= k && sen2[l] == sen1[j]) {
        //     l--;
        //     j--;
        // }

        // return l < k;


        // Approach 2: Dequeue

        if(sentence1.size() < sentence2.size()) {
            swap(sentence1, sentence2);
        }

        deque<string> deq1, deq2;

        stringstream ss1(sentence1);
        string token;

        while(ss1 >> token) {
            deq1.push_back(token);
        }

        stringstream ss2(sentence2);
        while(ss2 >> token) {
            deq2.push_back(token);
        }

        while(!deq1.empty() && !deq2.empty() && deq1.front() == deq2.front()) {
            deq1.pop_front();
            deq2.pop_front();
        }

        while(!deq1.empty() && !deq2.empty() && deq1.back() == deq2.back()) {
            deq1.pop_back();
            deq2.pop_back();
        }

        return deq2.empty();

    }
};

File: 1813-sentence-similarity-iii/1813-sentence-similarity-iii.js
/**
 * @param {string} sentence1
 * @param {string} sentence2
 * @return {boolean}
 */
var areSentencesSimilar = function(sentence1, sentence2) {
    let words1 = sentence1.split(" ")
    let words2 = sentence2.split(" ")

    if(words1.length > words2.length) {
        [words1, words2] = [words2, words1]
    }

    let i = 0;

    while(i < words1.length && i < words2.length && words1[i] == words2[i]) i++;

    let j = words1.length - 1
    let k = words2.length - 1

    while(j >= i && words1[j] == words2[k]) {
        j--;
        k--;
    }

     return j < i;
};